### PR TITLE
Function: isVehicleWheelCollided

### PR DESCRIPTION
**Pull request by ZReC**
**This function detect if car wheel is on air or in collision.**

Example:
`isVehicleWheelCollided ( theVehicle, "rear_left" )`
or
`isVehicleWheelCollided ( theVehicle, 1 )`
or
`theVehicle.isWheelCollided ( "front_right" )`

The wheels are:
> "front_left" or 0
> "rear_left" or 1
> "front_right" or 2
> "rear_right" or 3

Bikes only have front / rear left wheel .

If someone has a better name for the function please comment 👍

### CODE REVIEW

File: Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp [hunk 1]
```diff
@@ -913,6 +915,20 @@ int CLuaVehicleDefs::GetVehicleWheelStates ( lua_State* luaVM )
...
+{
+    CClientVehicle* pVehicle = NULL; eWheels wheel;
+    CScriptArgReader argStream ( luaVM );
+    argStream.ReadUserData ( pVehicle );
+    argStream.ReadEnumStringOrNumber ( wheel );
+
+    if ( !argStream.HasErrors () )
+        lua_pushboolean ( luaVM, pVehicle->IsWheelCollided ( wheel ) );
+    else
+        lua_pushboolean ( luaVM, false );
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp]: `m_pScriptDebugging->LogCustom ( luaVM, argStream.GetFullErrorMessage() );` is missing here

File: Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp [hunk 2]
```diff
@@ -913,6 +915,20 @@ int CLuaVehicleDefs::GetVehicleWheelStates ( lua_State* luaVM )
     return 1;
 }
 
+int CLuaVehicleDefs::IsVehicleWheelCollided ( lua_State* luaVM )
+{
+    CClientVehicle* pVehicle = NULL; eWheels wheel;
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp]: Consider using `nullptr` instead of NULL

**Comment by CrosRoad95** [comment]
Next can you add function getVehicleWheelContactMaterial(vehicle,wheel) it can be used for detect if player drive over grass

File: Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp [hunk 3]
```diff
@@ -917,16 +917,17 @@ int CLuaVehicleDefs::GetVehicleWheelStates ( lua_State* luaVM )
...
 int CLuaVehicleDefs::IsVehicleWheelCollided ( lua_State* luaVM )
 {
-    CClientVehicle* pVehicle = NULL; eWheels wheel;
+    CClientVehicle* pVehicle = nullptr; eWheels wheel;
     CScriptArgReader argStream ( luaVM );
     argStream.ReadUserData ( pVehicle );
     argStream.ReadEnumStringOrNumber ( wheel );
 
     if ( !argStream.HasErrors () )
         lua_pushboolean ( luaVM, pVehicle->IsWheelCollided ( wheel ) );
```
Comment (Citizen01) [inline Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp]: You should ```return 1;``` right after that line, with that change.

File: Client/mods/deathmatch/logic/lua/CLuaFunctionParseHelpers.cpp [hunk 4]
```diff
@@ -492,6 +492,12 @@ IMPLEMENT_ENUM_BEGIN ( eCursorType )
     ADD_ENUM ( CURSORTYPE_SEG_SIZING,   "segment_sizing" )  // segment sizing cursor (note: not in use)
 IMPLEMENT_ENUM_END ( "cursor-type" )
 
+IMPLEMENT_ENUM_BEGIN ( eWheels )
+    ADD_ENUM ( FRONT_LEFT_WHEEL,    "front_left" )
+    ADD_ENUM ( REAR_LEFT_WHEEL,     "rear_left" )
+    ADD_ENUM ( FRONT_RIGHT_WHEEL,   "front_right" )
+    ADD_ENUM ( REAR_RIGHT_WHEEL,    "rear_right" )
+IMPLEMENT_ENUM_END ( "wheels" )
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/lua/CLuaFunctionParseHelpers.cpp]: call it `wheel-position` instead

File: Client/mods/deathmatch/logic/lua/CLuaFunctionParseHelpers.cpp [hunk 5]
```diff
@@ -492,6 +492,12 @@ IMPLEMENT_ENUM_BEGIN ( eCursorType )
     ADD_ENUM ( CURSORTYPE_SEG_SIZING,   "segment_sizing" )  // segment sizing cursor (note: not in use)
 IMPLEMENT_ENUM_END ( "cursor-type" )
 
+IMPLEMENT_ENUM_BEGIN ( eWheels )
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/lua/CLuaFunctionParseHelpers.cpp]: `eWheelPosition`

**Comment by qaisjp** [comment]
I suggest the function is called `doesVehicleWheelCollide` or `getVehicleWheelCollide`.

Also, perhaps we should instead only support text arguments or number arguments. I lean towards only supporting text arguments. The simpler and easier to understand, the better. This is up for discussion.

File: Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp (hunk 3 above)
Comment (ZReC) [reply Client/mods/deathmatch/logic/luadefs/CLuaVehicleDefs.cpp]: **Yes, I realized later**
(I'm stupid..)

File: Client/mods/deathmatch/logic/lua/CLuaFunctionParseHelpers.cpp (hunk 5 above)
Comment (ZReC) [reply Client/mods/deathmatch/logic/lua/CLuaFunctionParseHelpers.cpp]: Should I change the label of the enum?

Comment (qaisjp) [reply Client/mods/deathmatch/logic/lua/CLuaFunctionParseHelpers.cpp]: Yeah

**Comment by jushar** [comment]
Could you please refactor `CVehicleSA::IsWheelCollided` to use the appropriate member attributes of `CVehicleSAInterface` instead of manual offset calculations.

E.g.:
```cpp
bool CVehicleSA::IsWheelCollided(std::uint8_t wheelPosition) // eWheelPosition might also be better here
{
  CVehicleSAInterface* vehicle = GetInterface();
  switch (vehicle->wheelType) // or whatever this is exactly
  {
  case 0:
    if (wheelPosition < 4)
      return vehicle->wheelCollisionState[wheelPosition] == 4.0f;
  // ...
}
```

**Comment by ZReC** [comment]
Not a bad idea, I'll do my best.

**Comment by CrosRoad95** [comment]
what if vehicle dont have 4 wheel? ex 3 ( dodo ) or motorbike ?

**Comment by ZReC** [comment]
In vehicles of 3 wheels are combined 2 in 1, in motorbikes only the left

**Comment by ZReC** [comment]
Jusonex, I have problems implementing the code to CVehicleSAInterface, it turns out that from offset 0x5A0 (1440 decimal) CVehicleSAInterface should give rise to a class more specific of the type of vehicle in particular (CAutomobile, CBike, CHeli, CTrain, CPlane) But instead the class extends with variables of different types of vehicles as of trains:
```
//1440
unsigned char m_ucTrackNodeID;  // Current node on train tracks
```

In my opinion would have to be rewritten this sector of the code, to make it more structured and manageable, meanwhile could graft the code of the collision but would not be the most appropriate.
**What do you think?**

Note: the offset 0x590 if it is the vehicle type, type uint8_t and its values correspond to:
**0 = car - plane | 5 = boat | 6 = train | 9 = bike**

**Comment by jushar** [comment]
> In my opinion would have to be rewritten this sector of the code, to make it more structured and manageable

Yes, that's the way to go - not as a part of this PR though.

I'd say update as much as possible (e.g. 0x590) and leave the rest as it is.

**Comment by ZReC** [comment]
Updated. Excuse me for the lack of activity, studies are killing me 😆

**Pull request merged**